### ISSUE TITLE
How to insure tracking from a specific point in time or startheight

### ISSUE THREAD

**Issue by bazooka70** [dgarage/NBXplorer#15]
I am experimenting with NBXplorer, and noticed a bit of an issue, which  I will (try to) explain.
NBXplorer server running and all is well.
So I created a `ExtPubKey ` which I intend to Track.

```
    public static void TestNBXplorer()
        {
            Network network = Network.TestNet;

            string m = "delay into trade deal exile effort badge install wheel despair attitude bike";
            Mnemonic mnemonic = new Mnemonic(m);
            ExtKey extMasterKey = mnemonic.DeriveExtKey();
            ExtPubKey extPubKey = extMasterKey.Neuter();

            ExtKey extKey0 = extMasterKey.Derive(new KeyPath("0/0"));
            Key privateKey = extKey0.PrivateKey;
            BitcoinSecret bitcoinPrivateKey = privateKey.GetWif(network);
            BitcoinPubKeyAddress publicAddress = bitcoinPrivateKey.GetAddress();
            Console.WriteLine(publicAddress); // mnB46n5rMSJqKTojJZHfyjab6rwajfoqho

            var xnetwork = new NBXplorerNetworkProvider(ChainType.Test).GetBTC();
            ExplorerClient client = new ExplorerClient(xnetwork);
            DerivationStrategyOptions derivationStrategyOptions = new DerivationStrategyOptions
            {
                Legacy = true // Use non-segwit
            };
            var userDerivationScheme = xnetwork.DerivationStrategyFactory.CreateDirectDerivationStrategy(extPubKey, derivationStrategyOptions);
            var _publicAddress = userDerivationScheme.Derive(new KeyPath("0/0")).ScriptPubKey.GetDestinationAddress(network);
            Console.WriteLine(_publicAddress); // mnB46n5rMSJqKTojJZHfyjab6rwajfoqho

            // client.Track(userDerivationScheme); // *** NOTE: this line was initially commented
            var utxos = client.GetUTXOs(userDerivationScheme, null, false);
            var coins = utxos.GetUnspentCoins();
            Console.WriteLine(coins.Length); 
            foreach (var coin in coins.OrderByDescending(x => x.Amount))
            {
                Console.WriteLine(coin.Amount.ToUnit(MoneyUnit.BTC));
            }
        }
```

I **deliberately** commented the `client.Track(userDerivationScheme); ` so I know NBXplorer won't track it.
I sent 2 transaction funds  to the above public adress `mnB46n5rMSJqKTojJZHfyjab6rwajfoqho`.
As expected NBXplorer client did not report any coins.

While the NBXplorer server was running, I Uncommented the `client.Track(userDerivationScheme);`,  and run my client programs again, I then sent another funds transaction to the address.

To my surprise (or not), NBXplorer reported **only the last transaction**, and did not see the 2 previous one.

When I **restarted** the NBXplorer server, it reported all the 3 transactions!
```
info: Events:         BTC: Money received in tpubD6NzVb...2G5ppdsMWmY-[legacy] i
n transaction 1a48fd...901045 (Unconfirmed)
info: Events:         BTC: Money received in tpubD6NzVb...2G5ppdsMWmY-[legacy] i
n transaction 6e611c...8e0259 (Unconfirmed)
info: Events:         BTC: Money received in tpubD6NzVb...2G5ppdsMWmY-[legacy] i
n transaction 9d8032...67e430 (Unconfirmed)

```
And so did the client.

I know the docs say:

```
Important Note
This tool will only start scanning from the configured startheight. (By default, the height of the blockchain during your first run) This mean that you might not see old payment from you HD key.

If you need to see old payments, you need to configure startheight to a specific height of your choice, then run with again with -rescan.
```

Now my question is how do I track an HD address, that already has funds in it before I first run NBXplorer server?
in other words, How do I tell the NBXplorer to start scanning from say 1.Jan.2018 (i don't know what the startheight block was then) ? and if I do specify a `startheight` do I always need to `rescan` when NBXplorer server starts?

I would have expected a command such as `client.RescanAndTrack(userDerivationScheme);` so I wont leave trasactions behinds. 

What is the correct strategy in this case so i won't need to rescan every time and make the chain.dat large enough to not skip transactions from a particular point in time?
I hope the question/issue is clear.

**Comment by bazooka70** [comment]
I changes `btc.startheight=1283670` and `btc.rescan=1`, after a few minutes NBXplorer finished rescanning.
I then removed the `btc.rescan=1` and restarted the NBXplorer. But It started again at `BTC: Starting scan at block 1287174` instead of  1283670.

**Comment by NicolasDorier** [comment]
So it can't rescan automatically. The reason is that while it is scanning it stop processing new blocks until it is done.

I plan one day to allow periodic scan simultaneously. I don't think it is that hard to add support for this.

The reason you saw the three transactions when you restarted is simple: When you restart, NBXplorer is sending a `mempool` message to the node. And the node will send you all he has in the mempool.

If a block were mind between after the two first transaction but before the third, then you would not have seen it.

```
I changes btc.startheight=1283670 (~11 days ago) and btc.rescan=1, after a few minutes NBXplorer finished rescanning.
I then removed the btc.rescan=1 and restarted the NBXplorer. But It started again at BTC: Starting scan at block 1287174 instead of 1283670.
```
This is expected behavior. The startheight is when the rescan will scan from if you ask for a rescan.

**Comment by NicolasDorier** [comment]
If the server is restarted you won't miss transactions. You can try.

Start the server at block X, make a transaction at X+1, verify it is detected.
Close the server.

At X+2 make a transaction and wait for it to confirm.

Restart the server.

You should see that the server pick up the transaction.

When you restart the server scan where it left off.
